README.md

[![Travis-CI Build
Status](https://travis-ci.org/hrbrmstr/udpprobe.svg?branch=master)](https://travis-ci.org/hrbrmstr/udpprobe)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/hrbrmstr/udpprobe?branch=master&svg=true)](https://ci.appveyor.com/project/hrbrmstr/udpprobe)
[![Coverage
Status](https://codecov.io/gh/hrbrmstr/udpprobe/branch/master/graph/badge.svg)](https://codecov.io/gh/hrbrmstr/udpprobe)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/udpprobe)](https://cran.r-project.org/package=udpprobe)

# udpprobe

Send User Datagram Protocol (‘UDP’) Probes and Receive Responses

## Description

Unfortunately (but understandably) R has no built in connection support
for the user datagram protocol (‘UDP’). We aim to somewhat fix that with
the ability to send a lightweight probe and receive lightweight
responses over ‘UDP’ to a target host/port. **Supports Windows, macOS
and Linux clients.**

## What’s Inside The Tin

*The following functions are implemented:*

The core probe function:

  - `udp_send_payload`: Send a UDP payload and return the response (if
    any)

Helpers for Ubiquity Discover Protocol probes:

  - `ubnt_discovery_probe`: Send a Ubiquiti Discovery probe to a host
    with a discovery query payload and receive a response udpprobe Send
    UDP payloads and gather response
  - `parse_ubnt_discovery_response`: Parser for Ubiquiti Discovery
    Protocol responses

Helpers for old, insecure internet services:

  - `udp_daytime`: Send an request for the day/time
  - `udp_echo`: Send an echo request to an echo server
  - `udp_qotd`: Send an request for the QOTD

## Installation

``` r
devtools::install_git("https://git.sr.ht/~hrbrmstr/udpprobe")
# or
devtools::install_gitlab("hrbrmstr/udpprobe")
# or
devtools::install_github("hrbrmstr/udpprobe")
```

## Usage

``` r
library(udpprobe)

# current version
packageVersion("udpprobe")
## [1] '0.2.1'
```

### Fun with manually crafting DNS records

What’s the IP address of `example.com`?

``` r
c(
  0xaa, 0xaa, # id
  0x01, 0x00, # |QR|   Opcode  |AA|TC|RD|RA|   Z    |   RCODE   |
  0x00, 0x01, # QDCOUNT num of questions
  0x00, 0x00, # ANCOUNT 0 since it's only for answers
  0x00, 0x00, # NSCOUNT 0 since it's only for answers
  0x00, 0x00, # ARCOUNT 0 sicne it's only for answers
  0x07,       # start of QNAME beginning with size of "example"
  0x65, 0x78, 0x61, 0x6d, 0x70, 0x6c, 0x65,  # example
  0x03,       # size of "com" (excluding null terminator)
  0x63, 0x6f, 0x6d, 0x00,  # "com" with null terminator
  0x00, 0x01, # QTYPE ("A" == 1)
  0x00, 0x01  # QCLASS (internet == 1)
) -> dns_req

(resp <- udp_send_payload("8.8.8.8", 53, dns_req))
##  [1] aa aa 81 80 00 01 00 01 00 00 00 00 07 65 78 61 6d 70
## [19] 6c 65 03 63 6f 6d 00 00 01 00 01 c0 0c 00 01 00 01 00
## [37] 00 4b 0c 00 04 5d b8 d8 22

paste0(as.integer(tail(resp, 4)), collapse = ".")
## [1] "93.184.216.34"

curl::nslookup("example.com")
## [1] "93.184.216.34"
```

### Live Ubiquiti Test (internet host redacted)

``` r
# the following just does:
#   udp_send_payload(Sys.getenv("UBNT_TEST_HOST"), 10001L, c(0x01, 0x00, 0x00, 0x00))
(x <- ubnt_discovery_probe(Sys.getenv("UBNT_TEST_HOST")))
## NULL

parse_ubnt_discovery_response(x)
## NULL
```

### Old, indecure internet services:

``` r
cat(udp_echo("91.207.125.246", "the message you want back"), "\n")
## the message you want back

cat(udp_qotd("119.48.167.199"), "\n")
## "When a stupid man is doing something he is ashamed of, he always declares
##  that it is his duty." George Bernard Shaw (1856-1950)

cat(udp_daytime("195.34.89.241"), "\n")
## 06 FEB 2019 16:19:04 CET
## 
```

## udpprobe Metrics

| Lang       | \# Files |  (%) | LoC |  (%) | Blank lines |  (%) | \# Lines |  (%) |
| :--------- | -------: | ---: | --: | ---: | ----------: | ---: | -------: | ---: |
| C          |        3 | 0.27 | 147 | 0.44 |          41 | 0.34 |        4 | 0.02 |
| R          |        6 | 0.55 | 146 | 0.43 |          38 | 0.31 |      140 | 0.70 |
| Rmd        |        1 | 0.09 |  30 | 0.09 |          36 | 0.30 |       57 | 0.28 |
| Dockerfile |        1 | 0.09 |  14 | 0.04 |           7 | 0.06 |        0 | 0.00 |

## Code of Conduct

Please note that this project is released with a [Contributor Code of
Conduct](CONDUCT.md). By participating in this project you agree to
abide by its terms.


NEWS.md
0.2.1
* Windows version supports dynamic allocation now
* Addded examples for some horribad old, insecure internet protocols

0.2.0
* Support for timeouts and large reponse sizes

0.1.0 
* Initial release


tests/test-all.R
library(testthat)
test_check("udpprobe")


tests/testthat/test-udpprobe.R
c(
  0xaa, 0xaa, # id
  0x01, 0x00, # |QR|   Opcode  |AA|TC|RD|RA|   Z    |   RCODE   |
  0x00, 0x01, # QDCOUNT num of questions
  0x00, 0x00, # ANCOUNT 0 since it's only for answers
  0x00, 0x00, # NSCOUNT 0 since it's only for answers
  0x00, 0x00, # ARCOUNT 0 sicne it's only for answers
  0x07,       # start of QNAME beginning with size of "example"
  0x65, 0x78, 0x61, 0x6d, 0x70, 0x6c, 0x65,  # example
  0x03,       # size of "com" (excluding null terminator)
  0x63, 0x6f, 0x6d, 0x00,  # "com" with null terminator
  0x00, 0x01, # QTYPE ("A" == 1)
  0x00, 0x01  # QCLASS (internet == 1)
) -> dns_req

context("packets move")

resp <- udp_send_payload("8.8.8.8", 53, dns_req)

expect_equal(
  paste0(as.integer(tail(resp, 4)), collapse = "."),
  curl::nslookup("example.com")
)

context("timeout works")

testthat::expect_warning(
  udp_send_payload("8.4.1.3", 55, dns_req, timeout=0.1)
)

context("echo works")
expect_equal(
  udp_echo("91.207.125.246", "the message you want back"),
  "the message you want back"
)

context("qotd works")
expect_is(udp_qotd("119.48.167.199"), "character")

context("daytime works")
expect_is(udp_daytime("195.34.89.241"), "character")


R/echo.R
#' Send an echo request to an echo server
#'
#' Unfortunately, `echo` servers are not great things to be running on the public
#' internet but you can stand one up locally if you don't know how to find one
#' externally (as of 2019 there are ~300K ov them on the public internet).
#'
#' @md
#' @param host host or IP address of target
#' @param payload what you want echoed back! Defaults to "`hello`".
#' @param port defaults to `7L` (the standard `echo` port)
#' @export
#' @family insecure old internet service utility functions
#' @return character vector
#' @examples \dontrun{
#' udp_echo("91.207.125.246", "the message you want back")
#' }
udp_echo <- function(host,  payload = "hello", port = 7L) {
  if (is.character(payload)) payload <- charToRaw(payload)
  res <- udp_send_payload(host, as.integer(port), payload, timeout = 5)
  readBin(res, "character") # in case there are nulls
}

#' Send an request for the QOTD
#'
#' Quote of the Day servers are also rly bad things to put
#' on the public internet but there are over 40K of them as
#' of 2019. Find an IP and make a query!
#'
#' @md
#' @param host host or IP address of target
#' @param port defaults to `17L` (the standard `echo` port)
#' @export
#' @family insecure old internet service utility functions
#' @return character vector
#' @examples \dontrun{
#' udp_qotd("119.48.167.199")
#' }
udp_qotd<- function(host,  port = 17L) {
  res <- udp_send_payload(host, as.integer(port), as.raw(0x00), timeout = 5)
  readBin(res, "character") # in case there are nulls
}

#' Send an request for the day/time
#'
#' `daytime` servers are also rly bad things to put
#' on the public internet but there are a few, still, of them as
#' of 2019. Find an IP and make a query!
#'
#' @md
#' @param host host or IP address of target
#' @param port defaults to `13L` (the standard `echo` port)
#' @export
#' @family insecure old internet service utility functions
#' @return character vector
#' @examples \dontrun{
#' udp_daytime("195.34.89.241")
#' }
udp_daytime<- function(host, port = 13L) {
  res <- udp_send_payload(host, as.integer(port), charToRaw("\n"), timeout = 5)
  readBin(res, "character") # in case there are nulls
}


src/main.c
#ifdef WIN64
#define IS_WINDOWS
#elif defined WIN32
#define IS_WINDOWS
#endif

#ifndef IS_WINDOWS
#include <Rinternals.h>
#include <Rdefines.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>

SEXP R_udp_send_payload(SEXP host, SEXP port, SEXP payload, SEXP timeout, SEXP buf_size) {

  int sockfd, n;
  socklen_t target_len;
  struct sockaddr_in target_addr;
  struct hostent *target;

  int sz = REAL(buf_size)[0];
  unsigned char *resp = (unsigned char *)R_alloc(sz, sizeof(unsigned char));

  if (resp == NULL) error("No memory available for UDP receive buffer.");

  struct timeval tv;
  float secs = REAL(timeout)[0];
  tv.tv_sec = (int)secs;
  tv.tv_usec = (secs - (int)secs) * 1000000;

  sockfd = socket(AF_INET, SOCK_DGRAM, 0);
  if (sockfd < 0) return(R_NilValue);

  if (setsockopt(sockfd, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv)) < 0) {
    Rf_warning("timeout setting failed");
    close(sockfd);
    return(R_NilValue);
  }

  target = gethostbyname(CHAR(STRING_ELT(host, 0)));
  if (target == NULL) {
    Rf_warning("gethostbyname failed");
    close(sockfd);
    return(R_NilValue);
  }

  bzero((char *)&target_addr, sizeof(target_addr));
  target_addr.sin_family = AF_INET;
  bcopy(
    (char *)target->h_addr,
	  (char *)&target_addr.sin_addr.s_addr,
	  target->h_length
	 );
  target_addr.sin_port = htons(INTEGER(port)[0]);
  target_len = sizeof(target_addr);

  n = sendto(sockfd, RAW(payload), LENGTH(payload), 0, (const struct sockaddr *)&target_addr, target_len);
  if (n <  0) {
    Rf_warning("socket sending failed");
    close(sockfd);
    return(R_NilValue);
  }

  bzero(resp, sz);
  n = recvfrom(sockfd, resp, sz, 0, (struct sockaddr *)&target_addr, &target_len);

  close(sockfd);

  if (n < 0) {
    Rf_warning("socket receiving failed");
    return(R_NilValue);
  }

  SEXP out = Rf_allocVector(RAWSXP, n);
  memcpy(RAW(out), resp, n);

  return(out);

}
#endif

src/main-win.c
#ifdef WIN64
#define IS_WINDOWS
#elif defined WIN32
#define IS_WINDOWS
#endif

#ifdef IS_WINDOWS
#include <Rinternals.h>
#include <Rdefines.h>
#include <stdlib.h>
#include <limits.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <winsock2.h>
#include <io.h>

#define BUFSIZE (1024*48)

SEXP R_udp_send_payload(SEXP host, SEXP port, SEXP payload, SEXP timeout, SEXP buf_size) {

  SOCKET sockfd;
  SOCKADDR_IN target_addr;
  WSADATA wsaData;

  int n, addr_size;

  int sz = REAL(buf_size)[0];
  unsigned char *resp = (unsigned char *)R_alloc(sz, sizeof(unsigned char));

  if (resp == NULL) error("No memory available for UDP receive buffer.");

  DWORD tv = REAL(timeout)[0] * 1000;

  WSAStartup(MAKEWORD(2, 2), &wsaData);

  sockfd = socket(PF_INET, SOCK_DGRAM, IPPROTO_UDP);
  if (sockfd == SOCKET_ERROR) {
    Rf_warning("invalid socket: %d", sockfd);
    WSACleanup();
    return(R_NilValue);
  }

  if (setsockopt(sockfd, SOL_SOCKET, SO_RCVTIMEO, (const char *)&tv, sizeof(tv))) {
    Rf_warning("timeout setting failed: %d", WSAGetLastError());
    closesocket(sockfd);
    WSACleanup();
    return(R_NilValue);
  }

  addr_size = sizeof(target_addr);

  memset(&target_addr, 0, addr_size);
  target_addr.sin_family = AF_INET;
  target_addr.sin_addr.s_addr = inet_addr(CHAR(STRING_ELT(host, 0)));
  target_addr.sin_port = htons(INTEGER(port)[0]);

  n = sendto(sockfd, (const char *)RAW(payload), LENGTH(payload), 0, (struct sockaddr *)&target_addr, sizeof(target_addr));
  if (n < 0) {
    Rf_warning("sendto failed: %d", WSAGetLastError());
    closesocket(sockfd);
    WSACleanup();
    return(R_NilValue);
  }

  memset(resp, 0, sz);
  n = recvfrom(sockfd, (char *)resp, sz, 0, (struct sockaddr *)&target_addr, &addr_size);

  if (n < 0) {
    Rf_warning("receive failed: %d", WSAGetLastError());
    closesocket(sockfd);
    WSACleanup();
    return(R_NilValue);
  }

  closesocket(sockfd);
  WSACleanup();

  SEXP out = Rf_allocVector(RAWSXP, n);
  memcpy(RAW(out), resp, n);

  return(out);

}
#endif

R/udp-payload.R
#' Send a UDP payload and return the response (if any)
#'
#' @note there's a hardcoded max response size of 4096 bytes for now #lazy
#' @param host host or ip address of target
#' @param port UDP port number
#' @param payload payload to send (raw bytes; [as.raw()] will be called if not)
#' @param timeout UDP socket transmission/reception timeout (seconds); Default is 5s.
#' @param max_response_size maximum size of response to accept (bytes). Default if 4096
#' @export
udp_send_payload <- function(host, port, payload, timeout = 5, max_response_size = 4096) {

  host <- host[1]
  port <- as.integer(port[1])
  max_response_size <- as.numeric(max_response_size)

  if (!is.raw(payload)) payload <- as.raw(payload)

  if (!grepl("^[[:digit:]\\.]$", host)) {
    host <- curl::nslookup(host, ipv4_only = TRUE, error = TRUE)
  }

  .Call(
    "R_udp_send_payload",
    host = host,
    port = port,
    payload = payload,
    timeout = timeout,
    tbuf_size = max_response_size
  )

}

R/ubnt-discovery.R
#' Send a Ubiquiti Discovery probe to a host with a discovery query payload and receive a response
#'
#' The `payload` is not fixed since it appears there may be other discovery
#' query payloads. Use [parse_ubnt_discovery_response()] to handle the the raw
#' result from this function.
#'
#' @md
#' @param host host or IP address of target
#' @param port defaults to `10001L` but folks may be running it on others
#' @param payload the discovery protocol query payload. The default is known to work.
#' @export
#' @return raw vector
#' @family Ubiquity Discovery Protocol functions
#' @examples \dontrun{
#' x <- ubnt_discovery_probe("some ip address or host name")
#' if (!is.null(x)) parse_ubnt_discovery_response(x)
#' }
ubnt_discovery_probe <- function(host, port = 10001L, payload = c(0x01, 0x00, 0x00, 0x00)) {
  udp_send_payload(host, as.integer(port), payload)
}

#' Parser for Ubiquiti Discovery Protocol responses
#'
#' Takes a raw or base64 encoded Ubiquity Discovery Protocol response and turns it into
#' a list.
#'
#' @md
#' @param a either a `raw` vector with the raw response byte data or a base64 encoded version of the same
#' @return an `ubi_d` classed `list`  with device `name`, `model_long` and `model_short` model info, `firmware` string
#'         station `esssid`, device `uptime` (if provided), device `macs` (MAC addresses) and
#'         `ips` (IPv4 addresses) and any unprocessed fields as named raw vectors (named by their field code).
#' @family Ubiquity Discovery Protocol functions
#' @export
#' @examples
#' paste0(
#'   c(
#'     "AQAAggIACiSkPH9XF8+QRUICAAokpDx+VxfAqAUBAQAGJKQ8flcXCgAEACylqg",
#'     "sADEFpclJvdXRlciBIUAwABkxBUC1IUA0AD1dDVGVsLVdpRmktNTcxNw4AAQMDA",
#'     "CJYTS5hcjcyNDAudjUuNS42LjE3NzYyLjEzMDUyOC4xNzU1EAAC5LI="
#'   ), collaspe = ""
#' ) -> base64_response
#'
#' parse_ubnt_discovery_response(base64_response)
#'
#' \dontrun{
#' x <- ubnt_discovery_probe("some ip address or host name")
#' if (!is.null(x)) parse_ubnt_discovery_response(x)
#' }
parse_ubnt_discovery_response <- function(a) {

  if (is.character(a)) a <- openssl::base64_decode(a)
  if (!is.raw(a)) return(NULL)

  if (!all(a[1:3] == c(0x01, 0x00, 0x00))) return(NULL)

  d <- list(
    name = character(0),
    model_long = character(0),
    model_short = character(0),
    firmware = character(0),
    essid = character(0),
    uptime = integer(0),
    macs = character(0),
    ips = character(0)
  )

  pos <- 5 # skip to the data

  repeat {

    field_type <- as.character(a[pos])
    field_len <- readBin(a[(pos+1):(pos+2)], "int", size=2, endian = "big")
    pos <- pos + 3

    if (field_len < 0) {
      warning("This appears to be a damaged response. Ignoring.", call. = FALSE)
      return(NULL)
    }

    if ((length(field_type)>0) && (field_len>0)) {
      switch(
        field_type,
        "01" = {
          d[["macs"]] <- c(d[["macs"]], paste0(as.character(a[pos:(pos+5)]), collapse=":")) # 6 bytes for MAC
        },
        "02" = {
          d[["macs"]] <- c(d[["macs"]], paste0(as.character(a[pos:(pos+5)]), collapse=":")) # 6 bytes for MAC
          d[["ips"]] <- c(d[["ips"]], paste0(as.integer(a[(pos+6):(pos+9)]), collapse=".")) # 4 byte IP address
        },
        "03" = {
          d[["firmware"]] <- readBin(a[pos:(pos+field_len-1)], "character") # field_len-length string (firmware info of device)
        },
        "0a" = {
          d[["uptime"]] <- readBin(a[pos:(pos+field_len-1)], "integer", size=4, endian = "big") # unsigned 32-bit integer uptime (seconds)
        },
        "0b" = {
          d[["name"]] <- readBin(a[pos:(pos+field_len-1)], "character") # field_len length-string (name of device)
        },
        "0c" = {
          d[["model_short"]] <- readBin(a[pos:(pos+field_len-1)], "character") # field_len-length string (short model name of device)
        },
        "0d" = {
          d[["essid"]] <- readBin(a[pos:(pos+field_len-1)], "character") # field_len-length string (essid of device)
        },
        "14" = {
          d[["model_long"]] <- readBin(a[pos:(pos+field_len-1)], "character") # field_len-length string (long model name of device)
        },
        # "known" unknown fields
        # "00"  field_len is usually 0 so this could be for padding but it's UDP so why?
        # "07" usually 1,408 bytes
        # "0e" usually 1 byte
        # "0f" usually 4 bytes
        # "10" usually 2 bytes
        # "16" tends to be 1,284 bytes of all 0x00's
        # "17" usually 4 byte but can be 0
        # "18" usually 1 byte
        # "2f" usually 8,192 bytes
        # "31" usually 13,056 bytes
        # "42" usually 16,707 bytes
        # "ec" usually 12,004 bytes but sometimes 2
        # "ff" usually 2 bytes
        d[[sprintf("ux%s", field_type)]] <- a[pos:(pos+field_len-1)] # unkown so raw
      )
    }

    pos <- pos + field_len
    if (pos > length(a)) break

  }

  if (length(d[["macs"]])) d[["macs"]] <- sort(unique(d[["macs"]]))
  if (length(d[["ips"]])) d[["ips"]] <- sort(unique(d[["ips"]]))

  class(d) <- c("ubnt_d")

  d

}

#' Printer for ubnt_d object
#'
#' @md
#' @param x ubnt_d object
#' @param ... ignored
#' @keywords internal
#' @export
print.ubnt_d <- function(x, ...) {

  mdl <- NULL
  if (length(x$model_short) > 0) {
    mdl <- sprintf("Model: %s", x$model_short)
  } else if (length(x$model_long) > 0) {
    mdl <- sprintf("Model: %s", x$model_long)
  }

  nm <- NULL
  if (length(x$name)) nm <- sprintf("Name: %s", x$name)

  fw <- NULL
  if (length(x$firmware)) nm <- sprintf("Firmware: %s", x$firmware)

  ut <- NULL
  if (length(x$uptime)) ut <- sprintf("Uptime: %s (hrs)", scales::comma_format(0.1)(x$uptime/60/60/24))

  cat("[", paste0(c(mdl, nm, fw, ut), collapse="; "), "\n", sep="")

}

